### PR TITLE
storage/mysql: ALTER SOURCE for MySQL Sources

### PR DESCRIPTION
**Pull request by rjobanp**
### Motivation

  * This PR adds a known-desirable feature.
Fixes https://github.com/MaterializeInc/materialize/issues/24975

Enables the `ALTER SOURCE .. ADD SUBSOURCE` statement and `DROP SOURCE <subsource>` for MySQL Sources, with equivalent semantics to the Postgres Source (except MySQL also supports the `IGNORE COLUMNS` option in addition to `TEXT COLUMNS`).

### Tips for reviewer

- Would recommend reviewing with whitespace removed.
- 2 of the commits are just refactors to allow the purification & sequencing of `ALTER SOURCE` statements to allow more than just Postgres sources.
- The `Implement purification for mysql alter source` commit has a large diff since I moved most of the mysql-specific subsource purification logic to a common method that can be shared between create & alter source.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/sql-parser/src/ast/defs/statement.rs [hunk 1]
```diff
@@ -2314,10 +2319,7 @@ impl WithOptionName for AlterSourceAddSubsourceOptionName {
     /// this value could contain sensitive user data. If you're uncertain, err
     /// on the conservative side and return `true`.
     fn redact_value(&self) -> bool {
-        match self {
-            AlterSourceAddSubsourceOptionName::Details
-            | AlterSourceAddSubsourceOptionName::TextColumns => false,
-        }
+        false
```
Comment (benesch) [inline src/sql-parser/src/ast/defs/statement.rs:2322]: Can you leave this as an explicit match? There's no guarantee that future options won't need redaction, and the explicit match forces all future additions to consider the question.

**Review by benesch — COMMENTED** [review]
Only looked at the parser changes on behalf of the SQL council.

File: test/mysql-cdc/alter-source.td [hunk 2]
```diff
@@ -8,7 +8,7 @@
 # by the Apache License, Version 2.0.
 
 $ set-sql-timeout duration=1s
-$ set-max-tries max-tries=20
+$ set-max-tries max-tries=14
```
Comment (nrainer-materialize) [inline test/mysql-cdc/alter-source.td:11]: Why do / can we decrease this value?

File: test/mysql-cdc/alter-source.td [hunk 3]
```diff
@@ -121,277 +121,321 @@ table_g               subsource
...
-# # CASCADE works
-# > DROP SOURCE table_e CASCADE;
-#
-# # IF NOT EXISTS + CASCADE works
-# > DROP SOURCE IF EXISTS table_e, table_f CASCADE;
-#
-# TEXT COLUMNS removed from table_f
-# > SELECT regexp_match(create_sql, 'TEXT COLUMNS = \((.*?)\)')[1] FROM (SHOW CREATE SOURCE mz_source);
-# <null>
+> DROP SOURCE table_a
```
Comment (nrainer-materialize) [inline test/mysql-cdc/alter-source.td:139]: Thanks for enabling and adding all these tests. Not much work left for me ;-)

File: test/mysql-cdc/alter-source.td (hunk 2 above)
Comment (rjobanp) [reply test/mysql-cdc/alter-source.td:11]: Turns out I can remove both of these overrides - the tests pass without them

**Review by benesch — COMMENTED** [review]
SQL parser changes LGTM from a SQL council perspective.

**Review by petrosagg — APPROVED** [review]
Changes look good with the only thing I'm not sure about is the thing I noted. Let's discuss in slack to make sure it's safe to do

File: src/sql/src/pure.rs [hunk 4]
```diff
@@ -941,207 +940,62 @@ async fn purify_create_source(
...
-                            }
-                            Ok((idents[0].as_str(), idents[1].as_str()))
-                        })
-                        .collect::<Result<Vec<_>, MySqlSourcePurificationError>>()?,
-                ),
-            };
+            // Retrieve the current @gtid_executed value of the server to mark as the effective
+            // initial snapshot point such that we can ensure consistency if the initial source
+            // snapshot is broken up over multiple points in time.
+            let initial_gtid_set =
```
[the remark this answers is not in the document]
Comment (rjobanp) [reply src/sql/src/pure.rs:946]: this is not changing the initial gtid set value! This function is `purify_create_source`, and I just extracted all the surrounding code into the mysql module `purify_subsources` function. In `purify_alter_source` down below we preserve this initial gtid set when constructing the new details
```
                // This value is not allowed to be altered in the source.
                initial_gtid_set: mysql_source_connection.details.initial_gtid_set,
```

[the remark this answers is not in the document]
Comment (petrosagg) [reply src/sql/src/pure.rs:946]: Ah, true! I missed that. 

>This effectively means that we may not be able to add a new subsource if the initial GTID set is no longer available in the replication binlog.

Yes, the current way of storing a static value won't cut it in the long run. I have an idea that should allow us to use the remap collection to track this frontier

**Pull request merged**